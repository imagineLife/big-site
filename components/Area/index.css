.explanation-wrapper{
	height: 50px;	
}

.explanatory-text{
	font-size: .75em;
}

rect.selection{
	stroke-dasharray: 4 12;
}