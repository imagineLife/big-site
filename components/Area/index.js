import WrappedArea from './State/ContextWrapper';

export default WrappedArea;
