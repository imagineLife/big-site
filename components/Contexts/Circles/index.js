import React from 'react';
import { getWordsByLength } from '../../lib/stats'
const CirclesContext = React.createContext()
const { Provider, Consumer } = CirclesContext

const CirclesProvider = (props) => {
 
 	//state
 	let [wordsByLength, setWordsByLength] = React.useState([]) 

 	//calculation && state-setter
 	const calcWordsByLength = (str) => {
 		setWordsByLength(getWordsByLength(str))
 	}
 	
	return(<Provider value={{calcWordsByLength, wordsByLength}}>
		{props.children}
	</Provider>)
}

export {
	CirclesProvider, 
	Consumer as CirclesConsumer, 
	CirclesContext
}