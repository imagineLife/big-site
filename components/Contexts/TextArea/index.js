import React, { useReducer, useEffect, createContext } from 'react';
import * as arr from 'd3-array';
import { getSentences } from '../../lib/stats';
import reducer from './reducer';
import textRes from './fullText.txt';
import themeRes from './themesArrSlim.json';

const TextAreaContext = createContext();
const { Provider, Consumer } = TextAreaContext;

const TextAreaProvider = (props) => {
  const initialContext = { text: '', curColor: null };

  const [textAnalyticsStore, textAreaDispatch] = useReducer(
    reducer,
    initialContext
  );

  /*
		load the text from textFile 'on load'
	*/
  useEffect(() => {
    // const textFileURL =
    // process.env.NODE_ENV == "development"
    // ?
    // "./../../fullText.txt"
    // : "./data/fullText.txt"
    // const themesURL = process.env.NODE_ENV == 'development' ? './../../themesArrSlim.json' : './data/themesArrSlim.json';

    // fetch(textFileURL).then(textFileRes =>
    //   textFileRes.text().then(textRes => {
    const sentences = getSentences(textRes);
    const maxWordCount = arr.max(sentences, (d) => d.wordCount);
    // fetch(themesURL).then(themeRes =>
    // themeRes.json().then(fetchedThemeData => {
    // update Provider state, triggering reducer with dispatched actions
    textAreaDispatch({
      type: 'DONE_FETCHING_TEXT',
      payload: {
        text: textRes,
        sentences,
        maxWords: maxWordCount,
        themes: themeRes, //fetchedThemeData,
      },
    });
    // })
    // )
    // })
    // )
  }, []);

  return (
    <Provider
      value={{
        textAreaDispatch,
        areaData: textAnalyticsStore.areaData,
        curColor: textAnalyticsStore.curColor,
        themesData: textAnalyticsStore.themes,
        ...textAnalyticsStore,
      }}
    >
      {props.children}
    </Provider>
  );
};

export { TextAreaProvider, Consumer as TextAreaConsumer, TextAreaContext };
