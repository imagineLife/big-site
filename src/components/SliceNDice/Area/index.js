import React from 'react';
import WrappedArea from './State/ContextWrapper';

export default WrappedArea;
