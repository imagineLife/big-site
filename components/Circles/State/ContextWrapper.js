import React from 'react';
import Circles from '..';
import { CirclesProvider } from './../../Contexts/Circles/'

const WrappedComponent = () => (
  <CirclesProvider>
    <Circles />
  </CirclesProvider>
);

export default WrappedComponent;
