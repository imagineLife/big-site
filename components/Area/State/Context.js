import React, { createContext, useContext, useReducer } from 'react';
import { TextAreaContext } from './../../Contexts/TextArea';

// STATE
import initialState from './reducer/initialState';
import reducer from './reducer';

const AreaContext = createContext();
const { Provider } = AreaContext;

const AreaProvider = ({ children, dims, hoverLine }) => {
  const { selectedAreaArr, sentences } = useContext(TextAreaContext);

  const [
    {
      curSentence,
      curSentenceObj,
      sentenceNumber,
      xOffset,
      showLine,
      offsetSentenceNumber,
    },
    dispatch,
  ] = useReducer(reducer, initialState);

  const stoppedMoving = () => {
    // @TODO: Payload is useless here
    dispatch({ type: 'STOP_MOVING', payload: true });
  };

  // mousedOver && mouseMove
  const moused = (d, xScale) => {
    const areaSVG = document.getElementsByClassName('area-svg')[0];

    // $FlowSVGBug
    const areaSVGXOffset = areaSVG.getBoundingClientRect().x;

    const xPos = d.pageX - areaSVGXOffset;

    if (xPos >= xScale.range()[0] - 5) {
      const thisSentence = Math.ceil(xScale.invert(xPos)) + selectedAreaArr[0];
      const sentenceWOffset = thisSentence - selectedAreaArr[0];
      if (thisSentence > -1) {
        dispatch({
          type: 'SET_CUR_SENTENCE',
          payload: {
            curSentenceObj: sentences[thisSentence],
            curSentence: sentences[thisSentence].text,
            sentenceNumber: thisSentence,
            offsetSentenceNumber: sentenceWOffset,
            showLine: true,
            xOffset: areaSVGXOffset,
          },
        });
      }
    }
  };

  return (
    <Provider
      value={{
        dims,
        hoverLine,
        curSentence,
        curSentenceObj,
        moused,
        offsetSentenceNumber,
        sentenceNumber,
        showLine,
        stoppedMoving,
        xOffset,
      }}
    >
      {children}
    </Provider>
  );
};
export { AreaContext, AreaProvider };
