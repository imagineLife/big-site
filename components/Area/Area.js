import React, { useContext, Fragment } from 'react';
// import './index.css';
import * as scale from 'd3-scale';
import * as d3Shape from 'd3-shape';
import { TextAreaContext } from './../Contexts/TextArea';
import { AreaContext } from './State/Context';

import { getSelectedSentences } from './helpers';

const Area = () => {
  const { sentences, selectedAreaArr, maxWordsPerSentence } =
    useContext(TextAreaContext);

  const {
    dims,
    hoverLine,
    curSentence,
    moused,
    offsetSentenceNumber,
    showLine,
    stoppedMoving,
  } = useContext(AreaContext);

  // sanity checking
  if (!sentences || !selectedAreaArr || !dims) {
    return <p>loading area...</p>;
  }

  // calc xScale
  const howManySentences = selectedAreaArr[1] - selectedAreaArr[0];
  const xScale = scale
    .scaleLinear()
    .domain([0, howManySentences])
    .range([0, dims.width]);

  // calc currently selected sentences
  const selectedSentences = getSelectedSentences({
    sentences,
    selectedAreaArr,
  });

  const yScale = scale
    .scaleLinear()
    .domain([0, maxWordsPerSentence]) // arr.max(selectedSentences, d => d.wordCount)])
    .range([dims.height, 0]);

  // build areaFn
  const areaFn = d3Shape
    .area()
    .x((d, i) => xScale(i))
    .y0(yScale(0))
    .y1((d) => yScale(d.wordCount))
    .curve(d3Shape.curveCatmullRom);

  // calc path
  const pathD = areaFn(selectedSentences);

  /*
    Hover-line
  */
  const scaledX = xScale(offsetSentenceNumber);

  const optHoverLine =
    !hoverLine ||
    offsetSentenceNumber < 0 ||
    offsetSentenceNumber > xScale.domain()[1] ||
    !showLine ? null : (
      <line
        pointerEvents="none"
        strokeWidth="1"
        stroke="rgb(150,150,150)"
        strokeDasharray="5 15"
        x1={scaledX}
        x2={scaledX}
        y1={yScale(0)}
        y2={yScale(
          Math.max(...selectedSentences.map((d) => d.wordCount)) * 1.05
        )}
      />
    );

  /*
    Hover-circle
  */
  const hoverCircle =
    !hoverLine || offsetSentenceNumber < 0 || !showLine ? null : (
      <circle
        pointerEvents="none"
        r={8}
        fill="rgba(255,255,255,.3)"
        stroke="white"
        strokeWidth="1"
        strokeDasharray="2 6"
        cx={scaledX}
        cy={yScale(selectedSentences[offsetSentenceNumber].wordCount)}
      />
    );

  return (
    <>
      <svg
        id="area"
        style={dims}
        className="area-svg"
        onMouseOver={(d) => moused(d, xScale)}
        onMouseMove={(d) => moused(d, xScale)}
        onMouseOut={stoppedMoving}
        onBlur={stoppedMoving}
      >
        <defs>
          <linearGradient id="areaGradient" gradientTransform="rotate(90)">
            <stop offset="1%" stopColor="rgb(147,147,147)" />
            <stop offset="95%" stopColor="rgba(147,147,147,.05)" />
          </linearGradient>
        </defs>

        {/* Area Path */}
        <path d={pathD} fill="rgba(147,147,147,.25)" />

        {/* Optional hovered-line */}
        {optHoverLine}

        {/* Optional hovered-circle */}
        {hoverCircle}
      </svg>
      <div className="explanation-wrapper">
        <p className="explanatory-text">
          {!curSentence &&
            'Hover over this area chart to highlight the sentence that was spoken at the specific point in time during the presidentâ€™s address.  HOVER'}
          {curSentence && curSentence}
        </p>
      </div>
    </>
  );
};

export default Area;
