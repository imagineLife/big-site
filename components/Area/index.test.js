import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import Area from './Area';

// state
import { TextAreaContext } from './../../components/Contexts/TextArea';
import { AreaContext } from './State/Context';

// mockData
import mockTextAreaContext from './mockData/text-area-state.json';
import mockAreaContext from './mockData/area-context-state.json';

describe('<Area/>', () => {
  const mousedFn = jest.fn();
  const textAreaState = {
    sentences: null,
    selectedAreaArr: [],
    maxWordsPerSentence: null,
  };

  const areaState = {
    dims: null,
    hoverLine: null,
    curSentence: null,
    moused: mousedFn,
    offsetSentenceNumber: null,
    showLine: null,
    stoppedMoving: null,
  };

  const compNoProps = mount(
    <TextAreaContext.Provider value={textAreaState}>
      <AreaContext.Provider value={areaState}>
        <Area />
      </AreaContext.Provider>
    </TextAreaContext.Provider>
  );
  it('matches snapshot', () => {
    expect(toJson(compNoProps)).toMatchSnapshot();
  });

  it('shows "loading area..." with no data', () => {
    expect(compNoProps.find('p').text()).toBe('loading area...');
  });

  describe('renders with mock state', () => {
    const mousedFn = jest.fn();
    const stopMovingFn = jest.fn();

    const areaWithProps = mount(
      <TextAreaContext.Provider value={mockTextAreaContext}>
        <AreaContext.Provider
          value={{
            ...mockAreaContext,
            moused: mousedFn,
            stoppedMoving: stopMovingFn,
          }}
        >
          <Area />
        </AreaContext.Provider>
      </TextAreaContext.Provider>
    );

    const areaSvg = areaWithProps.find('svg#area');

    it('renders area svg', () => {
      expect(areaSvg.length).toBe(1);
    });
    it('renders a path inside the svg', () => {
      expect(areaSvg.find('path').length).toBe(1);
    });

    describe('calls callback fns', () => {
      describe('calls moused', () => {
        it('onMouseOver', () => {
          areaSvg.simulate('mouseover');
          expect(mousedFn).toHaveBeenCalledTimes(1);
        });
        it('again onMouseMove', () => {
          areaSvg.simulate('mousemove');
          expect(mousedFn).toHaveBeenCalledTimes(2);
        });
      });
      describe('calls moused THEN calls stoppedMoving cb', () => {
        it('onMouseOver then onMouseOut', () => {
          areaSvg.simulate('mouseover');
          expect(mousedFn).toHaveBeenCalledTimes(3);
          areaSvg.simulate('mouseout');
          expect(stopMovingFn).toHaveBeenCalledTimes(1);
        });
      });
    });
  });
});
