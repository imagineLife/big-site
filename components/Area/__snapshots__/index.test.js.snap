// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Area/> matches snapshot 1`] = `
<Area>
  <p>
    loading area...
  </p>
</Area>
`;
